File: dialogio/pas/QuickSort.pas
// This file is part of Adlib Tracker II (AT2).
//
// SPDX-FileType: SOURCE
// SPDX-FileCopyrightText: 2014-2025 The Adlib Tracker 2 Authors
// SPDX-License-Identifier: GPL-3.0-or-later

// Local to `make_stream' procedure

procedure QuickSort(l,r: Word);

var
  i,j: Word;
  cmp: String;
  tmp: tSEARCH;

begin
  If (l >= r) then
    EXIT; //QuickSort
  cmp := stream.stuff[(l+r) DIV 2].name;
  i := l;
  j := r;

  Repeat
{$IFDEF GO32V2}
    While (i < r) and (stream.stuff[i].name < cmp) do
      Inc(i);
    While (j > l) and (stream.stuff[j].name > cmp) do
      Dec(j);
{$ELSE}
    While (i < r) and
          (CompareStr(stream.stuff[i].name,cmp) = isLess) do
      Inc(i);
    While (j > l) and
          (CompareStr(stream.stuff[j].name,cmp) = isMore) do
      Dec(j);
{$ENDIF}

    If (i <= j) then
      begin
        tmp := stream.stuff[i];
        stream.stuff[i] := stream.stuff[j];
        stream.stuff[j] := tmp;
        Inc(i);
        Dec(j);
      end;
  until (i > j);

  If (l < j) then QuickSort(l,j);
  If (i < r) then QuickSort(i,r);

  //EXIT //QuickSort
end;

File: dialogio/QuickSort.c
// This file is part of Adlib Tracker II (AT2).
//
// SPDX-FileType: SOURCE
// SPDX-FileCopyrightText: 2014-2025 The Adlib Tracker 2 Authors
// SPDX-License-Identifier: GPL-3.0-or-later

// Local to `make_stream' procedure

// l, r: 1..MAX_FILES
static void QuickSort (tSTREAM *stream, uint16_t l, uint16_t r)
{
  if (l < r)
  {
    uint16_t i = l, j = r;
    String s[FILENAME_SIZE + 1]; // as `name' member of `tSEARCH' structure

    CopyString (s, stream->stuff[(l + r) / 2 - 1].name, FILENAME_SIZE);

    do
    {
#if GO32
      while ((i < r) && (CompareStrings (stream->stuff[i - 1].name, s) < 0))
        i++;
      while ((j > l) && (CompareStrings (stream->stuff[j - 1].name, s) > 0))
        j--;
#else // !GO32
      while ((i < r) && (CompareStr (stream->stuff[i - 1].name, s) == isLess))
        i++;
      while ((j > l) && (CompareStr (stream->stuff[j - 1].name, s) == isMore))
        j--;
#endif // !GO32

      if (i <= j)
      {
        tSEARCH t;
        memcpy (&t, &stream->stuff[i - 1], sizeof (t));
        memcpy (&stream->stuff[i - 1], &stream->stuff[j - 1], sizeof (t));
        memcpy (&stream->stuff[j - 1], &t, sizeof (t));
        i++;
        j--;
      }
    } while (i <= j);

    if (l < j) QuickSort (stream, l, j);
    if (i < r) QuickSort (stream, i, r);
  }

  //EXIT //QuickSort
}
